cpp/src/Ice/IdleTimeoutTransceiverDecorator.cpp

```diff
@@ -0,0 +1,185 @@
+//
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+
+#include "IdleTimeoutTransceiverDecorator.h"
+#include "Ice/Buffer.h"
+
+#include <chrono>
+
+using namespace std;
+using namespace Ice;
+using namespace IceInternal;
+
+namespace
+{
+    class HeartbeatTimerTask final : public IceUtil::TimerTask
+    {
+    public:
+        HeartbeatTimerTask(const ConnectionIPtr& connection) : _connection(connection) {}
+
+        void runTimerTask() final
+        {
+            if (auto connection = _connection.lock())
+            {
+                connection->sendHeartbeat();
+            }
+            // else nothing to do, the connection is already gone.
+        }
+
+    private:
+        const std::weak_ptr<ConnectionI> _connection;
+    };
+
+    class IdleCheckTimerTask final : public IceUtil::TimerTask, public std::enable_shared_from_this<IdleCheckTimerTask>
+    {
+    public:
+        IdleCheckTimerTask(
+            const ConnectionIPtr& connection,
+            const chrono::milliseconds& idleTimeout,
+            const IceUtil::TimerPtr& timer)
+            : _connection(connection),
+              _idleTimeout(idleTimeout),
+              _timer(timer)
+        {
+        }
+
+        void runTimerTask() final
+        {
+            if (auto connection = _connection.lock())
+            {
+                if (connection->idleCheck())
+                {
+                    // Reschedule myself.
+                    // We set cancelPrevious to true in the unlikely event some other thread is concurrently reading and
+                    // already rescheduled this task.
+                    _timer->schedule(shared_from_this(), _idleTimeout, true);
+                }
+                // else the connection was aborted by the idle check or is no longer active for some other reason.
+            }
+            // else nothing to do, the connection is already gone.
+        }
+
+    private:
+        const std::weak_ptr<ConnectionI> _connection;
+        const chrono::milliseconds _idleTimeout;
+        const IceUtil::TimerPtr _timer;
+    };
+}
+
+void
+IdleTimeoutTransceiverDecorator::decoratorInit(const ConnectionIPtr& connection)
+{
+    _heartbeatTimerTask = make_shared<HeartbeatTimerTask>(connection);
+    if (_enableIdleCheck)
+    {
+        _idleCheckTimerTask = make_shared<IdleCheckTimerTask>(connection, _idleTimeout, _timer);
+    }
+}
+
+SocketOperation
+IdleTimeoutTransceiverDecorator::initialize(Buffer& readBuffer, Buffer& writeBuffer)
+{
+    SocketOperation op = _decoratee->initialize(readBuffer, writeBuffer);
```

Comment:
We can assert `_heartbeatTimerTask` is already set, to ensure decoratorInit was called before this.

Reply:
I don't think this makes sense. We know we always call init immediately after construction, because anybody else has a chance to use this transceiver.